library/src/main/java/kwasm/format/text/Tokenizer.kt

```diff
@@ -0,0 +1,135 @@
+ *
+ * From [the docs](https://webassembly.github.io/spec/core/text/lexical.html#tokens):
+ *
+ * The character stream in the source text is divided, from left to right, into a sequence of
+ * tokens, as defined by the following grammar.
+ *
+ * ```
+ *   token      ::= keyword | uN | sN | fN | string | id | ‘(’ | ‘)’ | reserved
+ *   keyword    ::= (‘a’ | … |‘z’) idchar* (if occurring as a literal terminal in the grammar)
+ *   reserved   ::= idchar+
+ * ```
+ *
+ * Tokens are formed from the input character stream according to the *longest match rule*. That is,
+ * the next token always consists of the longest possible sequence of characters that is recognized
+ * by the above lexical grammar. Tokens can be separated by white space, but except for strings,
+ * they cannot themselves contain whitespace.
+ */
+class Tokenizer {
+    /** Tokenizes source code from the provided [source] [Reader]. */
+    fun tokenize(
+        source: Reader,
+        context: ParseContext? = ParseContext("Unknown File", 1, 1)
+    ): List<Token> {
+        val rawTokens = source.use { Comment.stripComments(it.readText(), context) }.tokens
+        return rawTokens.flatMap(::buildActualTokensFrom)
+    }
+
+    /** Tokenizes source code from the provided [source] [String]. */
+    fun tokenize(
+        source: String,
+        context: ParseContext? = ParseContext("Unknown File", 1, 1)
+    ): List<Token> = tokenize(StringReader(source), context)
+
+    // TODO: it might be worth considering making a lot of these functions `inline fun`s.
+    private fun buildActualTokensFrom(token: RawToken): List<Token> = when {
+        token.sequence.isEmpty() -> emptyList()
+        Format.PATTERN.get().matchEntire(token.sequence) != null -> emptyList()
+        token.isKeyword() -> listOf(token.toKeyword())
+        token.isIntegerLiteral() -> listOf(token.toIntegerLiteral())
+        token.isFloatLiteral() -> listOf(token.toFloatLiteral())
+        token.isStringLiteral() -> listOf(token.toStringLiteral())
+        token.isIdentifier() -> listOf(token.toIdentifier())
+        token.isOpenParen() || token.isClosedParen() -> listOf(token.toParen())
+        token.isReserved() -> listOf(token.toReserved())
+        else -> {
+            val maxLengthTokenFind = listOf(
+                token.findKeyword(),
+                token.findIntegerLiteral(),
+                token.findFloatLiteral(),
+                token.findStringLiteral(),
+                token.findIdentifier(),
+                token.findParen(),
+                token.findReserved()
+            ).maxBy { it?.sequence?.length ?: -1 }.takeIf { it != null }
+                ?: throw ParseException(
+                    "No valid token found in sequence: \"${token.sequence}\"",
+                    token.context
+                )
+
+            // Recursively get token list from the contents before the longest found token.
+            val prefixTokens = buildActualTokensFrom(
```

Comment:
Could you explain this again, what cases do we need to recursively get tokens?

Reply:
An example could be:  `"this is a string"+1.5e10aKeyword`

The longest found token in this would be `"this is a string"`, but we still need to get `+1.5e10` and `aKeyword` out.

Reply:
I see